File: ios/SwiftKotlination/SwiftKotlinationTests/Sources/Features/TopStories/Model/Mocks/TopStoriesManagerMock.swift
import XCTest
@testable import SwiftKotlination

final class TopStoriesManagerMock {
    var result: Result<[Story], Error>
    var observer: Observer<[Story]>

    init(result: Result<[Story], Error> = .failure(NetworkError.invalidResponse)) {
        self.result = result
        self.observer = { _ in }
    }
}

extension TopStoriesManagerMock: TopStoriesManagerProtocol {
    @discardableResult
    func stories(_ observer: @escaping Observer<[Story]>) -> Disposable {
        self.observer = observer
        observer(result)

        return Disposable {}
    }

    func fetchStories() {
        stories(observer)
    }
}

File: android/SwiftKotlination/app/src/test/kotlin/fr/jhandguy/swiftkotlination/features/topstories/model/mocks/TopStoriesManagerMock.kt
package fr.jhandguy.swiftkotlination.features.topstories.model.mocks

import fr.jhandguy.swiftkotlination.features.topstories.model.TopStories
import fr.jhandguy.swiftkotlination.features.topstories.model.TopStoriesManagerInterface
import fr.jhandguy.swiftkotlination.observer.Disposable
import fr.jhandguy.swiftkotlination.observer.Observer
import fr.jhandguy.swiftkotlination.observer.Result

class TopStoriesManagerMock(var result: Result<TopStories>, var observer: Observer<TopStories> = {}) : TopStoriesManagerInterface {
    override suspend fun topStories(observer: Observer<TopStories>): Disposable {
        this.observer = observer
        observer(result)

        return Disposable {}
    }

    override suspend fun fetchStories() = observer(result)
}